### PR TITLE
Batch upgrade pipeline priority in pending queue

### PR DESCRIPTION
**Pull request by shuofan**
#### What type of this PR
feature

#### What this PR does / why we need it:
Batch upgrade pipeline priority in pending queue

#### Specified Reviewers:
/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |              |
| 🇨🇳 中文    |              |

### CODE REVIEW

File: apistructs/pipeline_priority.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package apistructs
+
+type PipelinePriorityUpgradeRequest struct {
+	PipelineIDs []uint64 `json:"pipelineIDs"`
```
Comment (sfwn) [inline apistructs/pipeline_priority.go:17]: 1,2,3 means 1 has highest priority, not 3,2,1.

Use a cleary field name.

File: modules/pipeline/services/queuemanage/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	for _, i := range queue.Usage.PendingDetails {
+		if i.Priority > maxPriority {
+			maxPriority = i.Priority
+		}
+	}
+
+	var priority = maxPriority
+	for _, pipeline := range pipelineIDs {
+		priority += 1
+		if err := reconciler.QueueManager.UpdatePipelinePriorityInQueue(strconv.FormatUint(queue.ID, 10), pipeline, priority); err != nil {
```
Comment (sfwn) [inline modules/pipeline/services/queuemanage/upgrade.go:42]: usee AddIntoQueue directly. It's idempotent.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: apistructs/pipeline_priority.go (hunk 1 above)
Comment (shuofan) [reply apistructs/pipeline_priority.go:18]: done

File: modules/pipeline/endpoints/queue_management.go [hunk 3]
```diff
@@ -176,3 +176,38 @@ func (e *Endpoints) deletePipelineQueue(ctx context.Context, r *http.Request, va
...
+	if len(req.PipelineIDs) == 0 {
+		return nil, apierrors.ErrUpgradePipelinePriority.InvalidParameter("pipelineIDs")
+	}
+
+	queue, err := e.queueManage.GetPipelineQueue(req.QueueID)
+	if err != nil {
+		return errorresp.ErrResp(err)
+	}
+
+	queue.Usage = e.reconciler.QueueManager.QueryQueueUsage(queue)
```
Comment (sfwn) [inline modules/pipeline/endpoints/queue_management.go:206]: why need query usage?

File: modules/pipeline/endpoints/queue_management.go [hunk 4]
```diff
@@ -176,3 +176,38 @@ func (e *Endpoints) deletePipelineQueue(ctx context.Context, r *http.Request, va
...
+
+	if err = e.queueManage.BatchUpgradePipelinePriority(queue, req.PipelineIDs, e.reconciler); err != nil {
```
Comment (sfwn) [inline modules/pipeline/endpoints/queue_management.go:208]: put BatchUpgradePipelinePriority into reconciler.queueManager.

e.queueManage is a simple service wrapper for db.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/spec/pipeline_extra.go [hunk 5]
```diff
@@ -115,8 +115,9 @@ type PipelineExtraInfo struct {
 }
 
 type QueueInfo struct {
-	QueueID        uint64 `json:"queueID"`
-	CustomPriority int64  `json:"customPriority"`
+	QueueID               uint64  `json:"queueID"`
+	CustomPriority        int64   `json:"customPriority"`
+	PriorityChangeHistory []int64 `json:"priorityChangeHistory,omitempty"`
```
Comment (sfwn) [inline modules/pipeline/spec/pipeline_extra.go:120]: add a comment or adjust field name to clarify original first or latest first.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/spec/pipeline_extra.go (hunk 5 above)
Comment (shuofan) [reply modules/pipeline/spec/pipeline_extra.go:121]: done

File: modules/pipeline/endpoints/queue_management.go (hunk 4 above)
Comment (shuofan) [reply modules/pipeline/endpoints/queue_management.go]: done

File: modules/pipeline/endpoints/queue_management.go (hunk 3 above)
Comment (shuofan) [reply modules/pipeline/endpoints/queue_management.go]: removed

File: modules/pipeline/pipengine/reconciler/queuemanage/types/manager.go [hunk 6]
```diff
@@ -24,4 +24,5 @@ type QueueManager interface {
 	QueryQueueUsage(pq *apistructs.PipelineQueue) *pb.QueueUsage
 	PutPipelineIntoQueue(pipelineID uint64) (popCh <-chan struct{}, needRetryIfErr bool, err error)
 	PopOutPipelineFromQueue(pipelineID uint64)
+	BatchUpdatePipelinePriorityInQueue(queueID uint64, pipelineID []uint64) error
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/types/manager.go:27]: - use `pq` as args uniformly
- use NewQueue(pq).ID() to calculate ID

File: modules/pipeline/pipengine/reconciler/queuemanage/types/queue.go [hunk 7]
```diff
@@ -29,4 +29,5 @@ type Queue interface {
 	RangePendingQueue()
 	AddPipelineIntoQueue(p *spec.Pipeline, doneCh chan struct{})
 	PopOutPipeline(p *spec.Pipeline)
+	UpdatePipelinePriority(p *spec.Pipeline, priority int64)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/types/queue.go:32]: - use mgr.`AddPipelineIntoQueue`
- `defaultQueue` add flag `pendingRangingQueue` when batch upgrade priorities, use `defer` to unset
- test with a large time sleep

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/reconciler/queuemanage/queue/add_pipeline.go [hunk 8]
```diff
@@ -25,7 +25,26 @@ import (
...
+	}
+	return pipelineID
+}
+
+// makeItemKey
+func makeItemKey(p *spec.Pipeline) string {
+	return strconv.FormatUint(p.ID, 10)
+}
 
+func (q *defaultQueue) AddPipelineIntoQueueUnblock(p *spec.Pipeline, doneCh chan struct{}) {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/queue/add_pipeline.go:47]: put this method after `makeItemKey`.

File: modules/pipeline/pipengine/reconciler/queuemanage/queue/add_pipeline.go [hunk 9]
```diff
@@ -49,13 +68,18 @@ func (q *defaultQueue) AddPipelineIntoQueue(p *spec.Pipeline, doneCh chan struct
 	//   else add to pending queue.
 	if p.Status.AfterPipelineQueue() {
 		q.eq.ProcessingQueue().Add(priorityqueue.NewItem(itemKey, priority, *createdTime))
-		go func() {
-			doneCh <- struct{}{}
-			close(doneCh)
-		}()
+		//    if doneCh is nil, external operations do not effect doneCh
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/queue/add_pipeline.go:71]: polish

File: modules/pipeline/pipengine/reconciler/queuemanage/queue/range_pending_queue.go [hunk 10]
```diff
@@ -59,7 +59,9 @@ func (q *defaultQueue) RangePendingQueue() {
 			q.unsetNeedReRangePendingQueueFlag()
 		}
 	}()
+	// TODO: query items every cycle instead of using original passed range, support items priority swap
 	q.eq.PendingQueue().Range(func(item priorityqueue.Item) (stopRange bool) {
+		time.Sleep(time.Hour * 1)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/queuemanage/queue/range_pending_queue.go:64]: ....

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**